### ISSUE TITLE
Memory leak when eviction is stressed?

### ISSUE THREAD

**Issue by cruftex** [ben-manes/caffeine#67]
Meanwhile I started experimenting on benchmarks that target at cache eviction operations.

I constructed this benchmark:

> Continuously access a cache of size 1M an ascending sequence, never producing any hit.
> Each thread starts at another offset.

Source:
https://github.com/cache2k/cache2k-benchmark/blob/master/jmh-suite/src/main/java/org/cache2k/benchmark/jmh/suite/eviction/symmetrical/NeverHitBenchmark.java

- JMH 1.11.3
- VM version: JDK 1.8.0_74, VM 25.74-b02
- VM options: -server -Xmx2G
- caffeine 2.2.7

The benchmark run expands to the maximum heap. After a full GC still 1.4 Gb is used.

A random sample with jmap during the benchmark iteration, shows:

````
$ jmap -histo:live ...

 num     #instances         #bytes  class name
----------------------------------------------
   1:      15486169      495557408  java.util.concurrent.ConcurrentHashMap$Node
   2:      15484002      495488064  com.github.benmanes.caffeine.cache.NodeFactory$SStMS
   3:      15484285      247748560  java.lang.Integer
   4:       5247787      125946888  com.github.benmanes.caffeine.SingleConsumerQueue$Node
   5:       5247786      125946864  com.github.benmanes.caffeine.cache.BoundedLocalCache$AddTask
   6:            16       67128576  [Ljava.util.concurrent.ConcurrentHashMap$Node;
   7:            77        8396088  [J
   8:          6103         490616  [C
...
````

I would have expected that the lagging eviction thread might be the cause, however, a benchmark run with 1,2 or 4 threads produces similar results. The memory isn't freeing up after the benchmark iteration is finished and two full GCs, as long as the cache reference is kept.

Ideas?

**Comment by ben-manes** [comment]
You're still trying to starve the out any other processes by inserting at a ridiculous rate. Its queuing theory and anyone doing this is using the wrong data structure, as it's not the purpose of a cache. You're not trying to get meaningful results or emulate anything real world. Adding a yield solves it.

**Comment by cruftex** [comment]
> I would have expected that the lagging eviction thread might be the cause, however, a benchmark run with 1,2 or 4 threads produces similar results.

**Comment by ben-manes** [comment]
Because the inserting thread is given full cpu quantums and run without any delays. The scheduler has no reason to deprioritze it. Eventually I'll add back pressure but it's been low priority given the artificial nature of the stress test. A yield to indicate real world work resolves the problem, indicating that it's not realistic.

**Comment by cruftex** [comment]
Hey, I am spending my time here, to help and give a useful bug report.

I added now for tear down:

````
    int cnt = 10;
    while (cnt-- >= 0) {
      Thread.sleep(3);
      Runtime.getRuntime().gc();
    }
````

GC says:

````
[Full GC (System.gc())  1620085K->1614786K(1864192K), 5.1165447 secs]
[Full GC (System.gc())  1616307K->1613851K(1864192K), 5.2649184 secs]
[Full GC (System.gc())  1615393K->1612797K(1864192K), 5.2099570 secs]
[Full GC (System.gc())  1614387K->1611609K(1864192K), 4.9768815 secs]
[Full GC (System.gc())  1612992K->1610564K(1864192K), 5.1135074 secs]
[Full GC (System.gc())  1611937K->1609454K(1864192K), 4.7591053 secs]
[Full GC (System.gc())  1610773K->1608395K(1864192K), 5.4222420 secs]
[Full GC (System.gc())  1609831K->1607246K(1864192K), 5.0065981 secs]
[Full GC (System.gc())  1608608K->1606134K(1864192K), 5.2420726 secs]
[Full GC (System.gc())  1607564K->1604988K(1864192K), 4.7984943 secs]
[Full GC (System.gc())  1606421K->1603833K(1864192K), 4.9631911 secs]
[Full GC (System.gc())  1605444K->1603829K(1864192K), 5.5147763 secs]
[Full GC (System.gc())  1604948K->1603769K(1864192K), 5.5836861 secs]
[Full GC (System.gc())  1604604K->1603659K(1864192K), 5.7016917 secs]
[Full GC (System.gc())  1604219K->1603563K(1864192K), 5.4341942 secs]
[Full GC (System.gc())  1603885K->1603536K(1864192K), 5.3091407 secs]
````

Besides, there is no change when I add `Thread.yield()` after every cache operation.

**Comment by ben-manes** [comment]
Sorry. I felt you threw it in my face last time instead of trying to be fair, explain, and analyze trade offs. But that's not a good reason to get annoyed, so I apologize.

I'll take a look. I'm fairly certain it's as explained above.

**Comment by drcrallen** [comment]
To make the prior statement a little clearer. The nodes having issues have about 4x the number of cores, and see about 7x the number of requests. The eviction count on both nodes reported by Caffeine is roughly equivalent.

**Comment by ben-manes** [comment]
You might try to capture the eviction weighted size using a `removal listener`. Unfortunately the weight isn't communicated so it would have to be re-weighed, but should good be a good temporary measure.

When I run through a [stress test](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/test/java/com/github/benmanes/caffeine/cache/Stresser.java) the GC stays fine and there isn't a stress on the total capacity. I do think backpressure here would be good, but it hasn't been necessary yet. A short-term solution is to have a rate limiter in front of writes to avoid a pile-up effect.

You might be promoting entries into old gen quicker than expected, since the cache does tend to hold long-term entries. You might also try using G1 (`-XX:+UseG1GC`) which does a much better job by collecting regions in parallel.

**Comment by cruftex** [comment]
The above mentioned stress test limits the key space to `1 << 20` and uses a pre-initialized array with integer instances. This limits already the amount of memory in the test.

I did the same in the above benchmark, but I had undesired effects that need evaluation and discussion another time. So I changed to a continuous ascending sequence to make sure that never a hit can happen. Now, there is a new key and value object created, which changes things dramatically.

**Comment by ben-manes** [comment]
Can't reproduce with my stress test with the ascending list. Note that I'm on a 4-core (8HT) laptop, and I think you said before you test on a 2-core laptop.

A bit of hacking during my commute using a chunked array queue keeps the bounding at a maximum of 1024 (16 initial) and processes 100M evictions / minute (1.67M/s). That's without any other system work except insertions with 16 threads. There's probably room to optimize that, such as a smarter back-off strategy, e.g. busy loop for half a context, try to schedule a drain, and yield.

I probably won't get time to work on a production quality version over the weekend. So it might be a little longer before I have the bandwidth to have back pressure baked into a release.

**Comment by ben-manes** [comment]
Oops, that should be 150M/m (2.55M/s). I forgot that it was unnecessarily using the removal listener for counting evictions, rather than enabling stats. That consumes some of the ForkJoinPool and caused context switches. Still not amazing, but should be acceptable.

**Comment by cruftex** [comment]
> ... I think you said before you test on a 2-core laptop.

Recently I switched to a dedicated benchmark system, which has 4 cores. The problem shows up also with 1 thread and with yields, so there should be enough CPU power available. Furthermore, if it is just a queuing problem, the memory should free up after the load ceases, which is not the case.

Besides the queuing rate problem, I still have the feeling that something else is odd here. Maybe it is not Caffeine at all, but a problem in the JDK, JMH or the OS scheduler?! One possibility is also that there is another cache implementation, bundled in the JMH jar, that screws things that Caffeine needs.

Maybe you can just take two minutes, pick my JMH benchmark and run it isolated within your environment?

**Comment by ben-manes** [comment]
Can you run with the snapshot? I'm about to check-in a fix using a bounded queue. I still need to play with the backoff behavior. Its a little naive, but seems okay for an initial commit.

**Comment by ben-manes** [comment]
What OS are you running? The Windows scheduler is known to be very quirky due to legacy reasons, so it doesn't give a fair balance across threads.

**Comment by cruftex** [comment]
> What OS are you running?

Oops, that's an essential information missing from my initial report. Sorry!

Ubuntu 14.04
Kernel:
Linux version 3.19.0-58-generic (buildd@lgw01-39) (gcc version 4.8.2 (Ubuntu 4.8.2-19ubuntu1) ) #64~14.04.1-Ubuntu SMP Fri Mar 18 19:05:43 UTC 2016

**Comment by cruftex** [comment]
> Can you run with the snapshot? I'm about to check-in a fix using a bounded queue. I still need to play with the backoff behavior. Its a little naive, but seems okay for an initial commit.

I can, and I am sure it will influence what we experience, but still I doubt that is the cause for the memory leak. 

Of course, limiting the queue is good, too. What would be the guarantee of the cache? Let's say if I configure the maximu size to 1000 entries, how many entries would the cache hold, regardless of the call rate?

**Comment by ben-manes** [comment]
The write buffer is a hand-off, meaning that there is a small amount of extra slack given and prioritized for immediate handling. A heavily overloaded system is already a problem in a clustered environment, as it won't be able to service requests with an acceptable latency. If the goal is keep latencies low and the variance narrow, then removing single points of contention is important. Hence most systems, e.g. async I/O, are used. That leads to queues, which are bounded to have back pressure if the system is falling over.

For my system the slack (write buffer's  maximum size) is 1,024. That showed a reasonable balance of handling a high write rates by applying a decent sized batch, without being excessive.

The alternatives are of course a coarse lock (wide variance on lock times) or handoff-with-notification (e.g. flat combining). Those synchronous modes trade off more unpredictable latencies to provide a more stricter constraint on a resource. That's not usually desirable in a server workload, hence the focus on latencies as a critical metric.

**Comment by cruftex** [comment]
Ha, I found the problem in my above testing scheme:

````
    int cnt = 10;
    while (cnt-- >= 0) {
      Thread.sleep(3);
      Runtime.getRuntime().gc();
    }
````

For this heap size the GC took about 5s (see above), so actually the sleep is not giving Caffeine CPU time. When there are just sleeps the memory goes away again.

**Comment by ben-manes** [comment]
Interesting.

You're running with Parallel GC? That's the default and is a throughput collector, so it prefers to STW. That can have pretty long pauses, but for batch jobs its a net win. I usually run using `-server -Xmx2G -XX:+UseG1GC -XX:-UseBiasedLocking` to try and reduce pause times, without over tuning the settings.

**Comment by cruftex** [comment]
> .... That leads to queues, which are bounded to have back pressure if the system is falling over.

For interactive applications minimizing the latency is, of course, a valid design choice. For batch processing (e.g. analytical workloads) you want throughput and trade in the low latency.

**Comment by cruftex** [comment]
> ... the system is falling over

For batch processing 100% CPU usage is the goal. This doesn't mean the system is "falling over" :)

**Comment by ben-manes** [comment]
Yep. I'm never claiming one design is better, as all are trade-offs. Finding a balance is of course hard, and resolving to better handle a variety of workloads takes time to mature. Other approaches may very well work wonderfully.

Maximizing cpu is definitely the goal of batch processing, but I/O usually starves it. Most caches are read centric and writes are very expensive, so some form of queuing exists there too (e.g. write-ahead logs). But that could have been the rub that I didn't account for as well, since druid batch centric. The bounded queue should help alleviate that now, without diminishing the primary goal of handling user-facing latencies.

**Comment by cruftex** [comment]
> You're running with Parallel GC?

At the moment I just use `-Xmx2G` and stay away from other tunings. For presenting benchmark results picking up "tuning parameters" is a quite sensible topic.

> I usually run using -server -Xmx2G -XX:+UseG1GC -XX:-UseBiasedLocking to try and reduce pause times, without over tuning the settings.

Thanks, will give it a try!

**Comment by ben-manes** [comment]
Closing as I think we have it fixed from both ends. I'll probably have a release by the end of the week, but want to let the changes bake a little.

**Comment by ben-manes** [comment]
Released `2.3.0`.

@drcrallen Please let me know if this release alleviates the problem that you experienced.

**Comment by drcrallen** [comment]
@ben-manes Just to follow up on this. I've had two tests run internally. One intentional and the other unintentional.

On the intentional side, changing to G1GC (with some basic tunings) kept the problematic nodes from having issues and tucked in the long-tail of response times (yay!)

On the unintentional side, some of our dogfood machines (internal use only) using caffeine had OOME. These machines had "maximum weight" tuned to estimate the on-heap size of the data. The actual heap-dump size retained by caffeine turned out to be 99.89% of the configured maximum weight. So I call that a major good sign.

So for my specific related issue, I think it was GC tunings.

**Comment by ben-manes** [comment]
Thanks for the feedback. That makes a lot of sense and mimics my previous analysis and experience. G1GC tends to do a very good job at reducing latencies and collecting garbage promptly due to its multi-region layout. The change in 2.3 were positive, even if most likely synthetic, so there was still an overall win.

**Comment by ben-manes** [comment]
The problem appears to be due to [JDK-8078490: Missed submissions in ForkJoinPool](https://bugs.openjdk.java.net/browse/JDK-8078490) and unrelated to the queue implementation. For JDK releases 8u40-8u60 the executor may never execute a task. This results in the maintenance work being scheduled but never run, so either a memory leak (SCQ) or halting (WB) depending on the release.

The fix is to either use a different executor (e.g. `Runnable::run`) or upgrade to 8u65 or above.
